controller/src/main/java/org/jboss/as/controller/ModelControllerLock.java

```diff
@@ -102,12 +143,31 @@ protected synchronized boolean tryAcquire(int permit) {
         }
 
         @Override
-        protected synchronized boolean tryRelease(int permit) {
-            final Object value = permitHolder.get();
-            if (value == null) {
+        protected int tryAcquireShared(final int permit) {
+            if (permitHolder != null) {
+                return -1;
+            }
+            if (compareAndSetState(0, 1)) {
+                return 1;
+            } else {
```

Comment:
I'm thinking that the options here are:

(1) Go back to using synchronized, and fix this as a seperate issue
(2) Synchronize on permitHolder, which would basically end up being the same as #1, and I think I'd rather have permitHolder just be an int anyway, rather than an Integer etc.
(3) Subclass AQS, and provide methods to getValue() / setValue(long). We could bitmask the int to store the permit in the upper bits, allowing just the use of a single CAS.
(4) Change operationID to use 28(or some amount) bits instead of the current 32 and mask the count in the int state variable in AQS.